### ISSUE TITLE
Support for source: Box<&dyn Error>

### ISSUE THREAD

**Issue by CodeSandwich** [lovasoa/custom_error#20]
Until 1.6 this compiled:
```
custom_error! { pub MyError
    Dynamic { source: Box<&dyn Error> } = "dynamic",
}
```
But didn't provide non-`None` `.source`. Since 1.7 this fails to compile, because it tries to return non-`None` `.source`:
```
the size for values of type `(dyn std::error::Error + 'static)` cannot be known at compilation time
```
It tries to return `&Box<&dyn Error>` and fails, because `Error` is not implemented for `&Error`, so `Box<&dyn Error>` doesn't implement `Error` and can't be uses as such unless manually dereferenced.

It would be great if source could be a boxed error!

**Comment by lovasoa** [comment]
Before 1.6, `custom_error` returned None for all source types with type or lifetime parameters. This was done with a custom matcher for error types, which could not match all possible rust types. So I switched to a standard type matcher, and removed the limitation of returning the error from `source()` only when it doesn't have type parameters. 

Unfortunately, I don't think there is any way to check whether a certain type can be returned or not from a `macxro_rules` macro. 

So, I am not really sure how this could be fixed, please let me know if you have any idea.

**Comment by CodeSandwich** [comment]
There was a similar discussion on quick-error: https://github.com/tailhook/quick-error/pull/28 They had exactly the same problem and figured out that calling `.borrow()` inside `.source` will fix it. They didn't merge it because quick-check API allowed a walkaround, but I doubt it can be used in custom_error.

**Comment by lovasoa** [comment]
Thank you for the bug report, and for the hint @CodeSandwich ! It's fixed.

**Comment by lovasoa** [comment]
I published [v1.7.1](https://crates.io/crates/custom_error/1.7.1), which implements the fix, to crates.io

**Comment by CodeSandwich** [comment]
My jaw literally fell, I can't believe you've decided, implemented and published it so quickly, godspeed! Thank you!
